airflow/contrib/hooks/gcp_compute_hook.py

```diff
@@ -106,49 +106,158 @@ def stop_instance(self, project_id, zone, resource_id):
+
+        :param project_id: Google Cloud Platform project ID where the Compute Engine
+        instance exists
+        :type project_id: str
+        :param body: Instance template representation as object according to
+            https://cloud.google.com/compute/docs/reference/rest/v1/instanceTemplates
+        :type body: dict
+        :param request_id: Optional, unique request_id that you might add to achieve
+           full idempotence (for example when client call times out repeating the request
+           with the same request id will not create a new instance template again)
+           It should be in UUID format as defined in RFC 4122
+        :type request_id: str
+        :return: True if the operation succeeded
+        :rtype: bool
+        """
+        response = self.get_conn().instanceTemplates().insert(
+            project=project_id,
+            body=body,
+            requestId=request_id
+        ).execute(num_retries=NUM_RETRIES)
+        operation_name = response["name"]
+        return self._wait_for_operation_to_complete(project_id, operation_name)
+
+    def get_instance_group_manager(self, project_id, zone, resource_id):
+        """
+        Retrieves Instance Group Manager by project_id, zone and resource_id.
+
+        :param project_id: Google Cloud Platform project ID where the Compute Engine
+        Instance Group Manager exists
+        :type project_id: str
+        :param zone: Google Cloud Platform zone where the Instance Group Manager exists
+        :type zone: str
+        :param resource_id: Name of the Instance Group Manager
+        :type resource_id: str
+        :return: Instance group manager representation as object according to
+          https://cloud.google.com/compute/docs/reference/rest/beta/instanceGroupManagers
+        :rtype: dict
+        """
+        response = self.get_conn().instanceGroupManagers().get(
+            project=project_id,
+            zone=zone,
+            instanceGroupManager=resource_id
+        ).execute(num_retries=NUM_RETRIES)
+        return response
+
+    def patch_instance_group_manager(self, project_id, zone, resource_id,
+                                     body, request_id=None):
+        """
+        Patches Instance Group Manager with the specified body.
+
+        :param project_id: Google Cloud Platform project ID where the Compute Engine
+            Instance Group Manager exists
+        :type project_id: str
+        :param zone: Google Cloud Platform zone where the Instance Group Manager exists
+        :type zone: str
+        :param resource_id: Name of the Instance Group Manager
+        :type resource_id: str
+        :param body: Instance Group Manager representation as json-merge-patch object
+         according to
+         https://cloud.google.com/compute/docs/reference/rest/beta/instanceTemplates/patch
+        :type body: dict
+        :param request_id: Optional, unique request_id that you might add to achieve
```

Comment:
Can we default this to a DagID+DagRunID+OperatorID? That way we get idempotence by default

Reply:
Not really. It has to be UDID-format, so just concatenating those won't work (I tried it before and I got an error about UDID - that's why I know UDID is needed - I also have comment below specifying that it has to be UDID format :)

Reply:
I thought it's a good idea first but after trying to implement it, I am very reluctant to do it. Indeed udid3/udid5 are nice and useful, but I have a bit of trouble with determining what is the combination of unique input values that we should use to generate the UDID. First of all the description of request_id is a bit vague in terms of what is the scope of uniqueness check - is it per project? or per account? per instance? or maybe globally unique (hopefully not) ? I will check with Alek and ask him to clarify. But there is another - more important-  question: do we want to make it unique per "dag_id + dag run id + task_id" ? or dag_id + dag run id + task_id + instance id from which we run it for example? Or some other unique ids should be taken into account ? I can easily imagine for example test setup and production setup with the same dag_id and task_id running with the same DAG in the same project. Should they be unique calls or not ?  What happens if we want to test a new DAG and we perform multiple requests from the same instance ? I can imagine situation that we manually delete such created template and want to repeat it again with exactly the same parameters - for testing. I am not sure how the API behaves in this case. How do we "clear" such unique requests that we already made and want to repeat them?

There are far too many cases I see that generating request_id automatically might break. I would really leave it up to the user to generate such request ids if they feel they could use it. It's equally easy for them to join dag_id and task_id (and other identifiers) using udid3/udid5 as it is for us, and they can do it in the way that will fit best their use case. I think we have not enough information at the moment operator is called to generate it ourselves.